### PR TITLE
journald: make SystemKeepFree=/RuntimeKeepFree= accept percentages

### LINKED ISSUE #9548: Allow percent values for SystemKeepFree and the like
**Is your feature request related to a problem? Please describe.**
We've had multiple disks run full, because we tried to set a different percent value for `SystemKeepFree`. This happened, because the default configuration of this value is (according to the documentation)

    The first pair defaults to 10% and the second to 15% of the size of the respective file system,

[from here](https://www.freedesktop.org/software/systemd/man/journald.conf.html). But of course one cannot replicate this setting, but has to write down absolute size units instead.

**Describe the solution you'd like**
We would like to be able to set this to 21%.

**Describe alternatives you've considered**
We're currently computing what 21% of the absolute size would be in Ansible and are templating that into the configuration. But This just is hard to write and read and would be much better if we could just write down that we want JournalD to keep 21% of the disk free.

### PR DESCRIPTION
**Pull request by wxphaha**
So far SystemKeepFree= and RuntimeKeepFree= only accepted absolute byte
sizes. Whenever the file system size changes (or the administrator would
like to keep a fraction of the disk free) an absolute value becomes
either too large or too small.

More importantly, SystemKeepFree= is enforced against the f_bfree
(free blocks for privileged users, i.e. including the reserved blocks)
value returned by statvfs(). On a typical ext4 file system 5% of the
disk is reserved for root, which means that with the default
keep_free of 5% journald is effectively unable to guarantee any free
space from the user's point of view: writes keep succeeding until the
file system is entirely full, because the reserved blocks were already
counted as available for journald.

An absolute SystemKeepFree= is not a usable workaround: the reserved 5%
scales with the file system size, so a fixed byte size either over- or
under-shoots depending on the disk. Give the administrator a generic way
to request a fraction of the file system instead, e.g.
'SystemKeepFree=21%'.

The percentage syntax ('NN%') is stored verbatim (as permyriad) in the
JournalMetrics structure. The effective keep_free value is computed on
demand by the new journal_effective_keep_free() helper from the file
system size known at the time, instead of being converted to bytes
eagerly when the journal files are opened. This way a SIGHUP reload
that does not reopen the journal files still picks up the current file
system size, and the check in journal_file_allocate() that refuses to
grow the journal (E2BIG) uses the same effective value as the space
accounting in cache_space_refresh().

An explicitly configured percentage is not subject to the
KEEP_FREE_UPPER cap, unlike the auto-derived default, so that the whole
disk can be reserved if requested. 0% explicitly means "keep nothing
free" (it maps to keep_free == 0), while any other percentage
invalidates a previously configured absolute byte value (mutual
exclusion of the two fields). The conversion divides first and then
multiplies (fs_size / 10000 * permyriad), so the result is bounded by
fs_size and cannot overflow, even on very large file systems.

No behavior change unless a percentage value is used. The auto-derived
default (5%) and absolute sizes are unaffected. Also adds tests for the
new parser, including boundary values (0%, 100%, 101%, malformed input)
and the mutual exclusion of the two fields.

Fixes: https://github.com/systemd/systemd/issues/9548

### CODE REVIEW

File: src/journal/journald-config.c [hunk 1]
```diff
@@ -463,6 +464,60 @@ int config_parse_line_max(
...
+                return 0;
+        }
+
+        r = parse_permyriad(rvalue);
+        if (r >= 0) {
+                if (r == 0)
+                        log_syntax(unit, LOG_WARNING, filename, line, 0,
+                                   "Specifying '0%%' for %s means journald may use the entire file system, ignoring KeepFree= entirely.", lvalue);
+
+                m->keep_free_permyriad = r;
```
Comment (poettering) [inline src/journal/journald-config.c:498]: why not set keep_free = UINT64_MAX here?

File: src/libsystemd/sd-journal/journal-file.c [hunk 2]
```diff
@@ -4107,7 +4107,26 @@ static void journal_default_metrics(JournalMetrics *m, int fd, bool compact) {
...
+                 * verbatim. */
+                if (fs_size > 0) {
+                        /* Multiplication first, then division, to keep as much
+                         * precision as possible. fs_size comes from statvfs() and
+                         * is bounded by the file system type, so this cannot
+                         * overflow in practice; if it ever did (u64_multiply_safe
+                         * returns 0), fall back to the default rather than
+                         * silently disabling the limit. */
+                        uint64_t keep = u64_multiply_safe(m->keep_free_permyriad, fs_size) / 10000;
+                        m->keep_free = keep > 0 ? PAGE_ALIGN_U64(keep) : DEFAULT_KEEP_FREE;
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-file.c]: i think it would be better to use MUL_SAFE() here

File: src/libsystemd/sd-journal/journal-file.c [hunk 3]
```diff
@@ -4107,7 +4107,26 @@ static void journal_default_metrics(JournalMetrics *m, int fd, bool compact) {
...
+                } else
+                        m->keep_free = DEFAULT_KEEP_FREE;
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-file.c]: i'd prefer if this calculation would be kept in a helper func that calculates the effective keep_free value, by looking at the permyriad, and the fixed value

File: src/libsystemd/sd-journal/journal-file.c [hunk 4]
```diff
@@ -4107,7 +4107,26 @@ static void journal_default_metrics(JournalMetrics *m, int fd, bool compact) {
...
+
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-file.c]: i am not too happy with propagating this into keep_free here.

i'd really prefer if we'd have a function manager_effective_keep_free() or so that calculcates the effective keep_free either by using the explicit absolute value, or by calculating it from the relative value.

in particular as it appears to me we should revalidatethis on on cache_space_refresh(), no?

**Comment by poettering** [comment]
conceptually looks fine, but some comments on the implementation, see above

File: src/journal/journald-config.c [hunk 5]
```diff
@@ -463,6 +464,60 @@ int config_parse_line_max(
...
+                m->keep_free = UINT64_MAX;
+                m->keep_free_permyriad = 0;
+                return 0;
+        }
+
+        r = parse_permyriad(rvalue);
+        if (r >= 0) {
+                if (r == 0)
+                        log_syntax(unit, LOG_WARNING, filename, line, 0,
+                                   "Specifying '0%%' for %s means journald may use the entire file system, ignoring KeepFree= entirely.", lvalue);
```
[the remark this answers is not in the document]
Comment (wxphaha) [reply src/journal/journald-config.c:496]: Fixed. The parser now sets both fields, covering both cases:

File: src/journal/journald-config.c (hunk 1 above)
Comment (wxphaha) [reply src/journal/journald-config.c:498]: Thank you for review.  Fixed. The parser now sets both fields.  
m->keep_free = r == 0 ? 0 : UINT64_MAX;

File: src/journal/test-journald-config.c [hunk 6]
```diff
@@ -160,4 +160,52 @@ TEST(config_forward_to_socket) {
...
+
+TEST(config_keep_free) {
+        /* Absolute sizes */
+        keep_free_parse_check("1K", 1024, 0);
+        keep_free_parse_check("1M", 1024 * 1024, 0);
+        keep_free_parse_check("1G", 1024 * 1024 * 1024, 0);
+
+        /* Percentages are stored verbatim as permyriad, to be converted
+         * to bytes once the file system size is known. Boundary values. */
+        keep_free_parse_check("0%", 0, 0);
```
[the remark this answers is not in the document]
Comment (wxphaha) [reply src/journal/test-journald-config.c:178]: Done. `keep_free_parse_check()` now starts from `JournalMetrics m = { .keep_free = UINT64_MAX }`
instead of a zeroed struct, so `keep_free_parse_check("0%", 0, 0)` actually fails unless the
parser sets `keep_free = 0` (a zeroed initial value would have made the assertion trivially
true and would not have caught the missing assignment).

File: src/libsystemd/sd-journal/journal-file.c [hunk 7]
```diff
@@ -4123,10 +4121,32 @@ static void journal_default_metrics(JournalMetrics *m, int fd, bool compact) {
...
 }
 
+uint64_t journal_effective_keep_free(const JournalMetrics *m, uint64_t fs_size) {
+        assert(m);
+
+        if (m->keep_free_permyriad > 0) {
+                /* An explicit percentage overrides keep_free, and is not
+                 * subject to the KEEP_FREE_UPPER cap. */
+                if (fs_size > 0)
+                        return MAX(PAGE_ALIGN_U64(fs_size / 10000 * m->keep_free_permyriad), (uint64_t) 1);
```
[the remark this answers is not in the document]
Comment (wxphaha) [reply src/libsystemd/sd-journal/journal-file.c:4135]: Done.  change the commit message

File: src/libsystemd/sd-journal/journal-file.c (hunk 2 above)
Comment (wxphaha) [reply src/libsystemd/sd-journal/journal-file.c]: Fixed.  Since multiplying before dividing could cause an overflow, the order has been changed to divide first and then multiply.

File: src/libsystemd/sd-journal/journal-file.c (hunk 3 above)
Comment (wxphaha) [reply src/libsystemd/sd-journal/journal-file.c]: Thank you for review.  Fixed.  i add new helper func journal_effective_keep_free

File: src/libsystemd/sd-journal/journal-file.c (hunk 4 above)
Comment (wxphaha) [reply src/libsystemd/sd-journal/journal-file.c]: Thank you for review.  Fixed.  i add new helper func journal_effective_keep_free

File: src/libsystemd/sd-journal/journal-file.c [hunk 8]
```diff
@@ -4123,10 +4123,46 @@ static void journal_default_metrics(JournalMetrics *m, int fd, bool compact) {
...
+         * the effective value to track that. */
+
+        if (m->keep_free_permyriad > 0) {
+                uint64_t keep;
+
+                /* The configured percentage is not subject to the
+                 * KEEP_FREE_UPPER cap: an explicit percentage is an
+                 * explicit request that must be honored verbatim. */
+                if (fs_size > 0 && MUL_SAFE(&keep, m->keep_free_permyriad, fs_size))
+                        return MAX(PAGE_ALIGN_U64(keep / 10000), (uint64_t) 1);
```
[the remark this answers is not in the document]
Comment (wxphaha) [reply src/libsystemd/sd-journal/journal-file.c]: Fixed. Since multiplying before dividing could cause an overflow, the order has been changed to divide first and then multiply.
